ts/ui/lazy/LazyHandler.ts

```diff
@@ -0,0 +1,483 @@
+   */
+  protected items: Map<string, LazyMathItem<N, T, D>> = new Map();
+
+  /**
+   * Add a MathItem to the list and return its ID
+   *
+   * @param {LazyMathItem} math   The item to add
+   * @return {string}             The id for the newly added item
+   */
+  public add(math: LazyMathItem<N, T, D>): string {
+    const id = String(this.id++);
+    this.items.set(id,  math);
+    return id;
+  }
+
+  /**
+   * Get the MathItem with the given ID
+   *
+   * @param {string} id       The ID of the MathItem to get
+   * @return {LazyMathItem}   The MathItem having that ID (if any)
+   */
+  public get(id: string): LazyMathItem<N, T, D> {
+    return this.items.get(id);
+  }
+
+  /**
+   * Remove an item from the map
+   *
+   * @param {string} id   The ID of the MathItem to remove
+   */
+  public delete(id: string) {
+    return this.items.delete(id);
+  }
+
+}
+
+/*==========================================================================*/
+
+/**
+ * The attribute to use for the ID on the marker node
+ */
+export const LAZYID = 'data-mjx-lazy';
+
+/**
+ * The properties added to MathItem for lazy typesetting
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ */
+export interface LazyMathItem<N, T, D> extends MathItem<N, T, D> {
+
+  /**
+   * True when the MathItem needs to be lazy compiled
+   */
+  lazyCompile: boolean;
+
+  /**
+   * True when the MathItem needs to be lazy displayed
+   */
+  lazyTypeset: boolean;
+
+  /**
+   * The DOM node used to mark the location of the math to be lazy typeset
+   */
+  lazyMarker: N;
+
+  /**
+   * The state to use when rerednering the math item (COMPILED vs TYPESET)
+   */
+  lazyState: number;
+
+  /**
+   * True if this item is a TeX MathItem
+   */
+  lazyTex: boolean;
+
+}
+
+/**
+ * The mixin for adding laxy typesetting to MathItems
+ *
+ * @param {B} BaseMathItem      The MathItem class to be extended
+ * @return {AssistiveMathItem}  The augmented MathItem class
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ * @template B  The MathItem class to extend
+ */
+export function LazyMathItemMixin<N, T, D, B extends Constructor<HTMLMathItem<N, T, D>>>(
+  BaseMathItem: B
+): Constructor<LazyMathItem<N, T, D>> & B {
+
+  return class extends BaseMathItem {
+
+    public lazyCompile: boolean = true;
+    public lazyTypeset: boolean = true;
+    public lazyMarker: N;
+    public lazyState: number = STATE.COMPILED;
+    public lazyTex: boolean = false;
+
+    /**
+     * Initially don't compile math, just use an empty math item,
+     *   then when the math comes into view (or is before something
+     *   that comes into view), compile it properly and mark the item
+     *   as only needing to be typeset.
+     *
+     * @override
+     */
+    public compile(document: LazyMathDocument<N, T, D>) {
+      if (this.lazyCompile) {
```

Comment:
The `if-if-else` is confusing. Could you turn it into a single guard:
```suggestion
      if (!this.lazyCompile || this.state() >= STATE.COMPILED) {
        super.compile(document);
￼        this.lazyState = STATE.TYPESET;
        return;
    }
```

Reply:
It's not the same thing.  In your setup, if `lazyCompile` is true but `this.state() >= STATE.COMPILED`, then `super.compile(document)` is called, but in mine it is not.  But I can do

```
      if (!this.lazyCompile) {
        super.compile(document);
        return;
      }
      if (this.state() < STATE.COMPILED) {
        this.lazyTex = (this.inputJax.name === 'TeX');
        this.root = document.mmlFactory.create('math');
        this.state(STATE.COMPILED);
      }
```

instead.